### PR TITLE
PT-166642114 Miner signalled consensus upgrade

### PR DESCRIPTION
**Pull request by lucafavatella**
In scope of https://www.pivotaltracker.com/story/show/166642114

TODO:
* [x] Validate this PR by design in the reference implementation.
  - See draft design at https://github.com/aeternity/aeternity/pull/2613
* [x] Review/confirm/refine/revise the value in the key block header field `info` being part of the validation rules of the new consensus protocol on the "signalling" block interval.

### CODE REVIEW

File: consensus/miner_signalled_consensus.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  decision globally the node may anyway taint the storage with blocks with
+  eventually wrong consensus protocol.
+
+*NOTE: This design choice may be lifted if problematic during the implementation
+ phase.*
+
+## Backwards Compatibility
+
+The signalling using the block header field `info` is backward compatible, as
+the meaning of the value of such field is not under consensus.
```
Comment (sennui) [inline consensus/miner_signalled_consensus.md]: `info` field value is under consensus, but the meaning of it - not. OK.

File: consensus/miner_signalled_consensus.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+The height at which the new consensus protocol may be active. Let us call it H2.
+
+H2 should cater for the time required by the node for aggregating the results of
+the "signalling" block interval.
+
+H2 must be strictly greater than HE.
+
+The difference between H2 and HE is recommended to be approximately two
+hours i.e. 40 (`= 2 * 20`).
+
```
Comment (sennui) [inline consensus/miner_signalled_consensus.md]: So we have: H1 < HS < HE < H2. OK.

File: consensus/miner_signalled_consensus.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+#### Majority
+
+The number of key blocks in the "signalling" block interval that is meant to
+activate the new consensus protocol.
+
+It must be positive.
+
+It is recommended to be 90% of (HE - HS). So it is 3024 (that is - 90%
+of 3360) if (HE - HS) is 3360.
```
Comment (jur0) [inline consensus/miner_signalled_consensus.md]: Is 90% going to be hardcoded or it's going be a configuration option?

Comment (lucafavatella) [reply consensus/miner_signalled_consensus.md]: Configuration option.

File: consensus/miner_signalled_consensus.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+The "signalling" block interval has a determined number of confirmations before
+the determined height at which the new consensus protocol may become
+active. This is in order to prevent the node from having to aggregate the
+results of the "signalling" block interval in a too short time interval.
+
+The value in the key block header field `info`:
+- Uses the whole field `info` i.e. is not a bitmask.
+- Is meant to be a pseudorandom number, so to limit the risk of the same value
+  being used by simultaneous competing consensus protocol proposals.
+- Is part of the validation rules of the new consensus protocol on the
```
Comment (lucafavatella) [inline consensus/miner_signalled_consensus.md]: This may need to be reviewed/confirmed/refined.

File: consensus/miner_signalled_consensus.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+The usage of the block header field `info` offers room for optimization for
+accommodating simultaneous hard-forking or soft-forking consensus proposals, by
+interpreting such `info` field as a bitmask rather than a value.
+
+### Soft-forking consensus proposals
+
+This proposal may be extended to soft-forking consensus proposals - as opposed
+to only hard-forking ones - by: relaxing the strict monotonicity of consensus
+versions; and excluding the value in the key block header field `info` in the
+"signalling" block interval from the validation rules of the new consensus
```
Comment (lucafavatella) [inline consensus/miner_signalled_consensus.md]: ... and this may be revised too.

File: consensus/miner_signalled_consensus.md [hunk 6]
```diff
@@ -186,5 +191,9 @@ then - with the same persisted storage - start (again) the node: the node
 needs not to consider any potential key block "signal" identified
 under the previous incorrect configuration.
 
+Any implementation of channels needs to ensure that the network
+reasonably settles on whether to activate the new consensus protocol
```
Comment (lucafavatella) [inline consensus/miner_signalled_consensus.md]: Planned as https://www.pivotaltracker.com/story/show/165214367

File: consensus/miner_signalled_consensus.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+per chain fork,
+the new consensus protocol becomes active at a determined
+height if at least a determined majority of the key blocks in a determined
+"signalling" block interval in such chain fork contains a determined value in
+the header field `info`.
+
+The "signalling" block interval has a determined number of confirmations before
+the determined height at which the new consensus protocol may become
+active. This is in order to prevent the node from having to aggregate the
+results of the "signalling" block interval in a too short time interval.
```
Comment (lucafavatella) [inline consensus/miner_signalled_consensus.md]: And also for preventing a user seeing the new protocol active after the fork height to stop node and configure unconditionally the new consensus protocol, and then restarting node (that had other chain fork X) and that chain fork gets prolonged.

File: consensus/miner_signalled_consensus.md [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+Any implementation needs to cater for the performance of the aggregation of the
+signalling e.g. slow multiple lookups on the storage.
+
+Any implementation needs to consider that the storage may contain blocks with
+the old and new consensus protocols at heights at which the new consensus
+protocol should have been applied. This is because the node should be aware of
+multiple chain forks and peers.
+
+Any implementation needs to be resilient to node restarts with
+accidental distinct miner signalling parameters in the configuration.
```
Comment (hanssv) [inline consensus/miner_signalled_consensus.md]: Since any hard fork will mean a new "release"/version of the node the _only_ configuration is whether to signal or not. This does not affect the history in any way. I.e. there is nothing that needs to be re-computed when you restart the node. The (partial) result of the vote could then be tallied in every block (not under consensus, but in the local database) as the node goes along - this means there is no need for an additional worker process and there is no need to have a time window after the vote period ends. 

I think this will simplify things considerably, but maybe I am missing something...

File: consensus/miner_signalled_consensus.md [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+Such extension should preferably be accompanied by a working implementation.
+
+## Implementation
+
+Any implementation needs to cater for the performance of the aggregation of the
+signalling e.g. slow multiple lookups on the storage.
+
+Any implementation needs to consider that the storage may contain blocks with
+the old and new consensus protocols at heights at which the new consensus
```
Comment (hanssv) [inline consensus/miner_signalled_consensus.md]: The node has _one_ chain at any point in time, there will not be multiple key-blocks at the same height, I don't understand this?!

Comment (jur0) [reply consensus/miner_signalled_consensus.md]:
```
                         [3] --- [4] --- [5] --- [6] --- [7] --- [8] ---  <- High difficulty
                          |
 [0] --- [1] --- [2] --- [3] --- [4] --- [5] --- [6] --- [7] --- [8] ---  <- Low difficulty
                  H1      HS            HE - 1    HE    H2 - 1    H2
```
There is just low difficulty chain at first and the fork signalling result is true (so switching to the new protocol at H2). Then there is a chain fork at height HS, this chain has higher difficulty, so it wins. The fork signalling result for high difficulty chain is false (so not switching to the new protocol at H2). Now there are 2 blocks in the db with different protocol versions at height H2 (but only the one in high difficulty chain is "active").

File: consensus/miner_signalled_consensus.md (hunk 8 above)
Comment (jur0) [reply consensus/miner_signalled_consensus.md]: The config should include more properties than "signal/not signal". There may be concurrent forks going on and we should make it possible to choose the signal value, etc.

If we wanted to keep the partial result in a local persistent db, then the result would apply just to one particular config. If we decide to restart the node, then the partial result might not be applicable anymore if we the config was changed. So we would need to recompute the result again and store it.

Comment (jur0) [reply consensus/miner_signalled_consensus.md]: > there is no need for an additional worker process and there is no need to have a time window after the vote period ends.

The time window after the fork signalling interval ends is there mainly because we want the chain between HS and HE to reasonably settle (so there is less likelihood of forks in the signalling interval and so the fork signalling result at H2 shouldn't change)

File: consensus/consensus.md [hunk 10]
```diff
@@ -170,6 +170,9 @@ EDDSA (curve 25519), signature is 64 bytes.
 
 The consensus protocol version number in each block is monotonically increasing in the chain of blocks.
 
+A means - not under consensus - of activating upgrades to the consensus protocol
+is [miner signalling](./miner_signalled_consensus.md).
```
Comment (ThomasArts) [inline consensus/consensus.md]: This sentence is hard to parse. Please clarify better

File: consensus/miner_signalled_consensus.md (hunk 8 above)
Comment (hanssv) [reply consensus/miner_signalled_consensus.md]: I don't think the signalling outcome should depend on the configuration... Either you decide to follow the outcome of the "poll" and then it is a feature that all nodes behave the same (since the poll is on chain they will) - if you don't want to follow the outcome you better run an older/other version of the node? **We** don't want the network to split, so let's not give them the tools for this ;-)

File: consensus/miner_signalled_consensus.md [hunk 11]
```diff
@@ -0,0 +1,200 @@
+# Miner Signalled Consensus Upgrade
+
+## Simple Summary
+
+This document describes the proposal of how nodes in the peer-to-peer network
```
Comment (ThomasArts) [inline consensus/miner_signalled_consensus.md]: --> This document describes a protocol to signal the readiness for, and activation of, a new hard-fork at a certain height.

File: consensus/miner_signalled_consensus.md [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+
+Activating a hard-forking new consensus protocol based on miner signalling
+prevents diverging the block chain in case a large fraction of the miners have
+not upgraded their nodes to a version able to follow the new consensus protocol.
+
+## Specification
+
+### Overview
+
+On a node configured for a miner-signalled consensus upgrade,
```
Comment (ThomasArts) [inline consensus/miner_signalled_consensus.md]: Nodes can be configured to handle a specific hard-fork as a reaction on signals from other nodes?

File: consensus/miner_signalled_consensus.md [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+per chain fork,
+the new consensus protocol becomes active at a determined
+height if at least a determined majority of the key blocks in a determined
+"signalling" block interval in such chain fork contains a determined value in
```
Comment (ThomasArts) [inline consensus/miner_signalled_consensus.md]: Explain the concept of "signalling block interval" first.

**Comment by ThomasArts** [comment]
The text is written by mixing implementation details and general concepts and vision. It would be nicer to first explain the why and concepts and then give how and implementation details.

File: consensus/miner_signalled_consensus.md [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+  unconditionally effective at the same height as the (past)
+  miner-signalled consensus upgrade height, the node may - depending
+  on the chain forks observed on the network - follow as best valid
+  chain a chain fork that upgraded to the new consensus protocol at
+  the configured height even if without a majority of "signals" in the
+  "signalling" block interval.
+
+### Parameters
+
+In order for the miner-signalled consensus protocol upgrade to succeed, each
```
Comment (ThomasArts) [inline consensus/miner_signalled_consensus.md]: If each node needs to have the same set, the this set better is under consensus

File: consensus/miner_signalled_consensus.md [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+The proposed new consensus protocol version.
+
+It must be strictly greater than C1 (strictly monotonically increasing consensus
+protocol version).
+
+#### Signal
+
+The "signal" value in the block header field `info`.
+
+It should be pseudorandom.
```
Comment (ThomasArts) [inline consensus/miner_signalled_consensus.md]: what does that mean?

File: consensus/miner_signalled_consensus.md [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+
+#### Signalling block interval start
```
Comment (ThomasArts) [inline consensus/miner_signalled_consensus.md]: better to start with this earlier in the story

File: consensus/miner_signalled_consensus.md (hunk 8 above)
Comment (jur0) [reply consensus/miner_signalled_consensus.md]: The configuration values would have to be hardcoded in the node then, and the only configuration option would be "follow signalling outcome/stay with the current protocol version".

I still prefer having the fork signalling configuration in the configuration file though. This is more flexible and it would allow the community to do protocol upgrade themselves (which may not be the best thing from our point of view). My understanding of fork signalling is that we just provide means of doing the protocol upgrades, we suggest some features and let the community decide. Maybe some other party can suggest new features as well, but they still will be able to use the fork signalling mechanism since it will be configurable.

This might be something we need to discuss further and more people should be involved...

File: consensus/miner_signalled_consensus.md (hunk 15 above)
Comment (jur0) [reply consensus/miner_signalled_consensus.md]: It's to make sure that in case of concurrent forks - meaning that aeternity will release a node which will try to upgrade to a new protocol, but also another party will do the same so the two signals will not be the same.

File: consensus/miner_signalled_consensus.md (hunk 12 above)
Comment (jur0) [reply consensus/miner_signalled_consensus.md]: It means that the fork signalling needs to be fork aware. It's showed in https://github.com/aeternity/protocol/pull/385#discussion_r325211457

So if the protocol upgrade happens on the current chain, it doesn't necessarily happen on all chain forks.

File: consensus/miner_signalled_consensus.md (hunk 14 above)
Comment (jur0) [reply consensus/miner_signalled_consensus.md]: Please have a look a the comments [here](https://github.com/aeternity/protocol/pull/385/files#r325690165).

File: consensus/miner_signalled_consensus.md [hunk 17]
```diff
@@ -0,0 +1,247 @@
+# Miner Signalled Consensus Upgrade
```
Comment (knarz) [inline consensus/miner_signalled_consensus.md]: In general I feel like this document is not so much about the miner signalling and more about how a node deals with the signals. It might help to clear that up?

File: consensus/miner_signalled_consensus.md [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+height). This is in order to:
+- Prevent the node from having to aggregate the results of the signalling
+  interval in a too short time interval.
+- Prevent a user from restarting the node and configuring the new consensus
+  protocol unconditionally when the signalling outcome is not confirmed with
+  reasonably high probability.
+
+The signal value is part of a block header in the signalling interval:
+- Is meant to be a pseudorandom number, so to limit the risk of the same value
+  being used by simultaneous competing consensus protocol proposals.
```
Comment (knarz) [inline consensus/miner_signalled_consensus.md]: What if someone chooses the same value intentionally?

File: consensus/miner_signalled_consensus.md [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+The height at which the signalling interval starts - included. Let us call it
+HS.
+
+HS should cater for the time required by most willing miners to adopt a version
+of the node able to signal and able to follow the new consensus protocol
+e.g. at least two weeks.
+
+HS must be strictly greater than H1.
+
+#### Signalling interval end
```
[the remark this answers is not in the document]
Comment (jur0) [reply consensus/miner_signalled_consensus.md]: The implementation doesn't allow anything to be configurable for mainnet and testnet. The only option is "follow miner signalling outcome" or "stay with the current protocol". Other environments allow configuration of all the signalling parameters.

File: consensus/miner_signalled_consensus.md (hunk 18 above)
Comment (jur0) [reply consensus/miner_signalled_consensus.md]: When the majority of the mining power signals a new protocol and then also follows it, then the nodes that were signalling the new protocol, but now are not able to follow it, will end up on a forked chain.

**Comment by dincho** [comment]
I'm afraid this is outdated.
